### DISCUSSION TITLE
Determine the notebook directory with zk's config file

### DISCUSSION THREAD

**Discussion by xlucn** [zk-org/zk-nvim#184] [Ideas]
Currently, zk-nvim will look for the default notebook directory according to `$ZK_NOTEBOOK_DIR`. However, zk provides another way with config section `[notebook]`.

Could the config file be used to determine the default notebook directory in zk-nvim, too?

**Comment by tjex** [comment]
zk-nvim is effectively a 'client' of sorts to `zk` / `zk`'s lsp. So the [notebook discovery](https://github.com/zk-org/zk-nvim?tab=readme-ov-file#notebook-directory-discovery) functionality of `zk` should work like you were executing zk commands from the cli. I would just recommend trying it out, and then ideally reporting back here with your findings so others can be informed too 👍

**Reply by xlucn** [reply]
In my case, `ZkNotes` fails when I didn't set `$ZK_NOTEBOOK_DIR` but set in config.toml
```
[notebook]
dir = "~/Code/test/zk/"
```

---

I tried something I can think of, although I am not sure what is the common way to debug this issue. All of these are tested in my `$HOME` folder.

1. Command line, zk successfully finds my notes with no issues.
   1. With `export ZK_NOTEBOOK_DIR="$HOME/Code/test/zk/"`

   ```
   $ export ZK_NOTEBOOK_DIR="$HOME/Code/test/zk/"
   $ zk list
   My first note vpg1-untitled.md (2 months ago)
   Found 1 note
   $ zk list $HOME
   My first note vpg1-untitled.md (2 months ago)
   Found 1 note
   ```

   2. With `unset ZK_NOTEBOOK_DIR`
   ```
   $ unset ZK_NOTEBOOK_DIR
   $ zk list
   My first note vpg1-untitled.md (2 months ago)
   Found 1 note
   $ zk list $HOME
   My first note vpg1-untitled.md (2 months ago)
   Found 1 note
   ```

2. In Neovim with zk-nvim's `ZkNotes`, behaviors start to differ.
   1. With `export ZK_NOTEBOOK_DIR="$HOME/Code/test/zk/"`, Telescope shows my notes.
   2. With `unset ZK_NOTEBOOK_DIR`, Error occurs:
   ```
   Error executing vim.schedule lua callback: /home/oliver/.local/share/nvim/lazy/zk-nvim/l
   ua/zk.lua:104: RPC[Error] code_name = InvalidRequest, message = "failed to open notebook
   : no notebook found in /home/oliver or a parent directory"                              
   stack traceback:                                                                        
           [C]: in function 'assert'                                                       
           /home/oliver/.local/share/nvim/lazy/zk-nvim/lua/zk.lua:104: in function 'handler
   '                                                                                       
           /usr/share/nvim/runtime/lua/vim/lsp/client.lua:687: in function ''              
           vim/_editor.lua: in function <vim/_editor.lua:0>                                
   ```

3. Some inspecting
   https://github.com/zk-org/zk-nvim/blob/dbf4eeab55b08856c9d6b6722dbff39630bb35eb/lua/zk/api.lua#L19-L25
   I tried to print the arguments:
   ```lua
   print(vim.inspect({ command = ..., arguments = ... }))
   ```
   1. With `export ZK_NOTEBOOK_DIR="$HOME/Code/test/zk/"`
   ```
   {
     arguments = { "/home/oliver/Code/test/zk/", {                                        
         select = { "title", "absPath", "path" }                                          
       } },                                                                               
     command = "zk.list"                                                                  
   }                                                                                      
   ```
   2. With `unset ZK_NOTEBOOK_DIR`
   ```
   {                                                                              
     arguments = { "/home/oliver", {                                                       
         select = { "title", "absPath", "path" }                                           
       } },                                                                                
     command = "zk.list"                                                                   
   }                                                                                       
   ```

   https://github.com/zk-org/zk-nvim/blob/dbf4eeab55b08856c9d6b6722dbff39630bb35eb/lua/zk.lua#L100-L107
   I tried to print something before the `assert`:
   ```lua
   assert(false, 'options:' .. vim.inspect(options) .. 'picker_options:' .. vim.inspect(picker_options) .. "Notes:" .. vim.inspect(notes))
   ```
   1. With `export ZK_NOTEBOOK_DIR="$HOME/Code/test/zk/"`
   ```
   options:{                                                              
     select = { "title", "absPath", "path" }                                              
   }picker_options:{                                                                      
     title = "Zk Notes"                                                                   
   }Notes:{ {                                                                             
       absPath = "/home/oliver/Code/test/zk/vpg1-untitled.md",                            
       path = "vpg1-untitled.md",                                                         
       title = "My first note"                                                            
     } }
   ```
   2. With `unset ZK_NOTEBOOK_DIR`
   ```
   options:{                                                                
     select = { "title", "absPath", "path" }                                               
   }picker_options:{                                                                       
     title = "Zk Notes"                                                                    
   }Notes:nil
   ```

Not sure if these helps.

**Comment by tjex** [comment]
Are you setting this:

```toml
[notebook]
dir = "~/Code/test/zk/"
```

In the _global_ config.toml or the local one?
It's also not clear where you're running these commands from btw. The below will work if you are within a `zk` notebook. But it should _not_ work if you are higher than that directory (as shown below in my test).

> With unset ZK_NOTEBOOK_DIR
> ```
> $ unset ZK_NOTEBOOK_DIR
> $ zk list
> My first note vpg1-untitled.md (2 months ago)
> Found 1 note
> $ zk list $HOME
> My first note vpg1-untitled.md (2 months ago)
> Found 1 note
> ```


The `zk list /any/random/path/without/a/notebook` command (while having `$ZK_NOTEBOOK_DIR` set) is in keeping with the [notebook discovery](https://github.com/zk-org/zk-nvim?tab=readme-ov-file#notebook-directory-discovery) logic. 

## My Tests

But, I tested on my end with a global config at `~/.config/zk/config.toml` (and commenting out my `$ZK_NOTEBOOK_DIR` variable) with my main zk notebook. So:

```toml
[notebook]
dir = "~/wikis/ps"
```

And everything works as expected.
I also tried your `export / unset` steps and it all worked for me.
(the below is scrollback from my cli, "base >" is my prompt, it's not shell redirection..) 

```text
~/.config
base > zk list
zk: error: failed to open notebook: no notebook found in /Users/tjex/.config or a parent directory
~/.config
base > export ZK_NOTEBOOK_DIR="$HOME/wikis/ps"
~/.config
base > echo $ZK_NOTEBOOK_DIR
/Users/tjex/wikis/ps
~/.config
base > zk list
2024-03-18 -> d/2024/2024-03-18.md ... (5 months ago)
2024-03-28 -> d/2024/2024-03-28.md ... (5 months ago)
...
Found 453 notes
~/.config
base > unset ZK_NOTEBOOK_DIR
~/.config
base > echo $ZK_NOTEBOOK_DIR

~/.config
base > zk list
zk: error: failed to open notebook: no notebook found in /Users/tjex/.config or a parent directory
```

So for the moment, I'm thinking there is some error in your config / testing process.

Be sure to try:

1. setting `$ZK_NOTEBOOK_DIR` in your `.zprofile` or `.zshrc`
2. opening a _new_ terminal window / instance
3. run `zk list`
4. 5. go to any random markdown document _outside_ of a `zk` notebook, run `:ZkNotes`

Then:

1. Remove `$ZK_NOTEBOOK_DIR` in your zsh file.
2. add your above toml entry to `~/.config/zk/config.toml`
3. open a new shell instance
5. run `zk list`
6. go to any random markdown document _outside_ of a `zk` notebook, run `:ZkNotes`

Both should work.

**Reply by xlucn** [reply]
Thanks for the reply!

I am putting the settings in the ~/.config/zk/config.toml. And all my tests are run in the $HOME folder with the settings.

My expectation is that zk-nvim finds the default notebook path no matter where I run nvim in, without the env var ZK_NOTEBOOK_DIR set, as you described at the end of your reply. But my tests shows otherwise.

**Reply by itsdalmo** [reply]
I'm having the same issue as @xlucn. `ZK_NOTEBOOK_DIR` did not seem to work for me either, at least not with a `~/` in there (got it working with an absolute path).

**Reply by xlucn** [reply]
> `ZK_NOTEBOOK_DIR` did not seem to work for me either, at least not with a `~/` in there (got it working with an absolute path).

@itsdalmo, I know that one. `~` is actually a trick that some shells (like bash) and some applications do, which will be replaced with user's `$HOME` folder. However, it is not am enforced standard. Zk, for example, doesn't convert `~` to your home folder and will look for a folder named literally `~`. Thus, you have to set that environment variable with something like `$HOME`, say
```sh
export ZK_NOTEBOOK_DIR="$HOME/foo/bar"
```

**Reply by tjex** [reply]
It's not that `zk` doesn't expand `~`, it's that tilde expansion doesn't work at all in shell environment variables.

```text
~/.config/nvim main* ⇡ ≡
base > echo $FOO
~/wikis

~/.config/nvim main* ⇡ ≡
base > cd $FOO
cd: no such file or directory: ~/wikis

~/.config/nvim main* ⇡ ≡
base > source ~/.zprofile         <--- after editing ~/.zprofile with export FOO="$HOME/wikis"

~/.config/nvim main* ⇡ ≡
base > echo $FOO
/Users/tjex/wikis

~/.config/nvim main* ⇡ ≡
base > cd $FOO

~/wikis
base >
```

For example you can use `~` in the `[notebook]` path (in the global config at `~/.config/zk/config.toml`, although I realise this is the point of concern with this issue).
